In [96]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [97]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_path = Path("Credit_Risk/lending_data.csv")
lending_data_df = pd.read_csv("lending_data.csv")

# Review the DataFrame
lending_data_df.head(5)

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [98]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data_df["loan_status"]

# Separate the X variable, the features
x = lending_data_df.drop(columns="loan_status")

In [99]:
# Review the y variable Series
y[0:11]

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
Name: loan_status, dtype: int64

In [100]:
# Review the X variable DataFrame
x.head(11)

loan_size  interest_rate  borrower_income  debt_to_income  \
0     10700.0          7.672            52800        0.431818   
1      8400.0          6.692            43600        0.311927   
2      9000.0          6.963            46100        0.349241   
3     10700.0          7.664            52700        0.430740   
4     10800.0          7.698            53000        0.433962   
5     10100.0          7.438            50600        0.407115   
6     10300.0          7.490            51100        0.412916   
7      8800.0          6.857            45100        0.334812   
8      9300.0          7.096            47400        0.367089   
9      9700.0          7.248            48800        0.385246   
10     9300.0          7.085            47300        0.365751   

    num_of_accounts  derogatory_marks  total_debt  
0                 5                 1       22800  
1                 3                 0       13600  
2                 3                 0       16100  
3                 5                 1       22700  
4                 5                 1       23000  
5                 4                 1       20600  
6                 4                 1       21100  
7                 3                 0       15100  
8                 3                 0       17400  
9                 4                 0       18800  
10                3                 0       17300

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [101]:
# Check the balance of our target values
lending_data_df["loan_status"].value_counts()

loan_status
0    75036
1     2500
Name: count, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [102]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
x_train, x_test, y_train, y_test = train_test_split(x,y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [103]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
log_regress_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model =log_regress_model.fit(x_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [104]:
# Make a prediction using the testing data
predict_testing_data = lr_model.predict(x_test)
predict_testing_data_df = pd.DataFrame({"Predictions": predict_testing_data, "Actual Result": y_test}).reset_index(drop=True)
predict_testing_data_df.head()

Predictions  Actual Result
0            0              0
1            0              0
2            0              0
3            0              0
4            0              0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [105]:
# Print the balanced_accuracy score of the model
bal_accu_score = print(balanced_accuracy_score(y_test, predict_testing_data))

0.9647722372908478


In [106]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, predict_testing_data)

array([[18662,    95],
       [   41,   586]], dtype=int64)

In [107]:
# Print the classification report for the model
class_report= classification_report(y_test, predict_testing_data)
print(class_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18757
           1       0.86      0.93      0.90       627

    accuracy                           0.99     19384
   macro avg       0.93      0.96      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistics regression model shows an accuracy percent at 99%. There is a difference in the precision from the healthy loans vs high risk loans. The healthy loans have a 100% precision whereas the high-risk loan has 86%. Based on this data, additional testing may need to occur to ensure that we are receiving a true accurate result. However based on the data provided, the accuracy of the model is showing to be a 96%. 

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [108]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
random_oversampler = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
x_resampled, y_resampled, = random_oversampler.fit_resample(x_train, y_train)

In [109]:
# Count the distinct values of the resampled labels data
y_resampled.value_counts()

loan_status
0    56279
1    56279
Name: count, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [110]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
random_os = LogisticRegression(solver='lbfgs', random_state=1)
random_os


# Fit the model using the resampled training data
random_os_model = random_os.fit(x_resampled, y_resampled)


# Make a prediction using the testing data
random_os_predict = random_os_model.predict(x_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [111]:
# Print the balanced_accuracy score of the model 
print(balanced_accuracy_score(y_test,random_os_predict))

0.9946754168714812


In [112]:
# Generate a confusion matrix for the model
confusion_matrix(y_test,random_os_predict)

array([[18647,   110],
       [    3,   624]], dtype=int64)

In [113]:
# Print the classification report for the model
print(classification_report(y_test, random_os_predict))

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18757
           1       0.85      1.00      0.92       627

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.96     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistics regression model shows an accuracy percent at 99%. There is a difference in the precision from the healthy loans vs high risk loans. The healthy loans have a 100% precision whereas the high-risk loan has 85%. Based on this data and the prediction models, the logistic regression model has an overall great accuracy for healthy loans and for high-risk loans while using the imbalanced data.  